File: sdk-go/memory_test.go
package boxxkite

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"testing"
)

func TestRemember_SendsContentScopeAndKind(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/memory" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["content"] != "The user prefers TypeScript." {
			t.Errorf("unexpected content: %v", body["content"])
		}
		if body["scope"] != "default" {
			t.Errorf("unexpected scope: %v", body["scope"])
		}
		if body["kind"] != "fact" {
			t.Errorf("unexpected kind: %v", body["kind"])
		}
		writeJSON(t, w, 201, `{"id": "mem-1", "content": "The user prefers TypeScript."}`)
	})
	defer closeServer()

	mem, err := client.Remember(context.Background(), RememberRequest{Content: "The user prefers TypeScript."})
	if err != nil {
		t.Fatalf("Remember: %v", err)
	}
	if mem.ID != "mem-1" {
		t.Errorf("unexpected id: %+v", mem)
	}
}

func TestIngestMemory_SendsContentAndScope(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/memory/ingest" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["content"] != "long transcript" {
			t.Errorf("unexpected content: %v", body["content"])
		}
		writeJSON(t, w, 201, `{"source_session_id": null, "memories": []}`)
	})
	defer closeServer()

	resp, err := client.IngestMemory(context.Background(), IngestMemoryRequest{Content: "long transcript"})
	if err != nil {
		t.Fatalf("IngestMemory: %v", err)
	}
	if len(resp.Memories) != 0 {
		t.Errorf("expected no memories, got: %+v", resp.Memories)
	}
}

func TestRecall_SendsQueryAndLimit(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/memory/search" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		if r.URL.Query().Get("q") != "TypeScript" {
			t.Errorf("unexpected q: %v", r.URL.Query().Get("q"))
		}
		if r.URL.Query().Get("limit") != "10" {
			t.Errorf("unexpected limit: %v", r.URL.Query().Get("limit"))
		}
		writeJSON(t, w, 200, `{"query": "TypeScript", "memories": []}`)
	})
	defer closeServer()

	resp, err := client.Recall(context.Background(), "TypeScript", RecallOptions{})
	if err != nil {
		t.Fatalf("Recall: %v", err)
	}
	if resp.Query != "TypeScript" {
		t.Errorf("unexpected query: %+v", resp)
	}
}

func TestMemoryProfile_ReturnsStaticAndDynamic(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/memory/profile" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		writeJSON(t, w, 200, `{"static": [], "dynamic": []}`)
	})
	defer closeServer()

	resp, err := client.MemoryProfile(context.Background(), "", 0)
	if err != nil {
		t.Fatalf("MemoryProfile: %v", err)
	}
	if len(resp.Static) != 0 || len(resp.Dynamic) != 0 {
		t.Errorf("unexpected profile: %+v", resp)
	}
}

func TestListMemories_ReturnsEmptySliceWhenNone(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet || r.URL.Path != "/v1/memory" {
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		writeJSON(t, w, 200, `[]`)
	})
	defer closeServer()

	memories, err := client.ListMemories(context.Background(), ListMemoriesOptions{})
	if err != nil {
		t.Fatalf("ListMemories: %v", err)
	}
	if memories == nil || len(memories) != 0 {
		t.Errorf("expected empty (non-nil) slice, got: %+v", memories)
	}
}

func TestGetMemory_ReturnsResult(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/memory/mem-1" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		writeJSON(t, w, 200, `{"id": "mem-1"}`)
	})
	defer closeServer()

	mem, err := client.GetMemory(context.Background(), "mem-1")
	if err != nil {
		t.Fatalf("GetMemory: %v", err)
	}
	if mem.ID != "mem-1" {
		t.Errorf("unexpected memory: %+v", mem)
	}
}

func TestMemoryRelations_ReturnsResult(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/memory/mem-1/relations" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		writeJSON(t, w, 200, `{"memory_id": "mem-1", "relations": []}`)
	})
	defer closeServer()

	resp, err := client.MemoryRelations(context.Background(), "mem-1", 0)
	if err != nil {
		t.Fatalf("MemoryRelations: %v", err)
	}
	if resp.MemoryID != "mem-1" {
		t.Errorf("unexpected response: %+v", resp)
	}
}

func TestForgetMemory(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodDelete || r.URL.Path != "/v1/memory/mem-1" {
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.WriteHeader(204)
	})
	defer closeServer()

	if err := client.ForgetMemory(context.Background(), "mem-1"); err != nil {
		t.Fatalf("ForgetMemory: %v", err)
	}
}

func TestExportMemories_ReturnsResult(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/memory/export" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		writeJSON(t, w, 200, `{"memories": [], "relations": []}`)
	})
	defer closeServer()

	resp, err := client.ExportMemories(context.Background(), "")
	if err != nil {
		t.Fatalf("ExportMemories: %v", err)
	}
	if resp == nil {
		t.Errorf("expected a non-nil export payload")
	}
}

func TestImportMemories_SendsMemoriesAndRelations(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/memory/import" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		memories, _ := body["memories"].([]any)
		if len(memories) != 1 {
			t.Errorf("unexpected memories: %v", body["memories"])
		}
		writeJSON(t, w, 201, `{"source_session_id": null, "memories": [{"id": "mem-1"}]}`)
	})
	defer closeServer()

	resp, err := client.ImportMemories(context.Background(), ImportMemoriesRequest{
		Memories: []map[string]any{{"content": "x"}},
	})
	if err != nil {
		t.Fatalf("ImportMemories: %v", err)
	}
	if len(resp.Memories) != 1 || resp.Memories[0].ID != "mem-1" {
		t.Errorf("unexpected response: %+v", resp)
	}
}

File: sdk-rust/tests/memory_test.rs
mod common;

use boxxkite_client::{IngestMemoryOptions, ListMemoriesOptions, RememberOptions};
use serde_json::json;
use wiremock::matchers::{body_json, method, path, query_param};
use wiremock::{Mock, ResponseTemplate};

#[tokio::test]
async fn remember_sends_content_scope_and_kind() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/memory"))
        .and(body_json(json!({
            "content": "The user prefers TypeScript.",
        })))
        .respond_with(ResponseTemplate::new(201).set_body_json(json!({
            "id": "mem_1",
            "scope": "default",
            "kind": "fact",
            "content": "The user prefers TypeScript.",
            "metadata": {},
            "source_session_id": null,
            "created_at": "2026-01-01T00:00:00Z",
            "updated_at": "2026-01-01T00:00:00Z",
            "expires_at": null,
            "source_content": null,
            "document_date": null,
            "event_dates": [],
            "importance": 0.5,
            "access_count": 0,
            "last_accessed_at": null,
            "superseded_at": null,
            "superseded_by_id": null
        })))
        .mount(&server)
        .await;

    let memory = client
        .remember("The user prefers TypeScript.", RememberOptions::new())
        .await
        .expect("remember should succeed");

    assert_eq!(memory.id, "mem_1");
}

#[tokio::test]
async fn ingest_memory_sends_content_and_scope() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/memory/ingest"))
        .and(body_json(json!({ "content": "long transcript" })))
        .respond_with(ResponseTemplate::new(201).set_body_json(json!({
            "source_session_id": null,
            "memories": []
        })))
        .mount(&server)
        .await;

    let resp = client
        .ingest_memory("long transcript", IngestMemoryOptions::new())
        .await
        .expect("ingest_memory should succeed");

    assert!(resp.memories.is_empty());
}

#[tokio::test]
async fn recall_sends_query_and_limit() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/memory/search"))
        .and(query_param("q", "TypeScript"))
        .and(query_param("limit", "10"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "query": "TypeScript",
            "memories": []
        })))
        .mount(&server)
        .await;

    let resp = client
        .recall("TypeScript", None, None)
        .await
        .expect("recall should succeed");

    assert_eq!(resp.query, "TypeScript");
}

#[tokio::test]
async fn memory_profile_returns_static_and_dynamic() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/memory/profile"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "static": [],
            "dynamic": []
        })))
        .mount(&server)
        .await;

    let resp = client
        .memory_profile(None, None)
        .await
        .expect("memory_profile should succeed");

    assert!(resp.r#static.is_empty());
    assert!(resp.dynamic.is_empty());
}

#[tokio::test]
async fn list_memories_returns_empty_vec_on_no_content() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/memory"))
        .respond_with(ResponseTemplate::new(200))
        .mount(&server)
        .await;

    let memories = client
        .list_memories(ListMemoriesOptions::default())
        .await
        .expect("list_memories should succeed");

    assert!(memories.is_empty());
}

#[tokio::test]
async fn get_memory_returns_the_memory() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/memory/mem_1"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "id": "mem_1",
            "scope": "default",
            "kind": "fact",
            "content": "x",
            "metadata": {},
            "source_session_id": null,
            "created_at": "2026-01-01T00:00:00Z",
            "updated_at": "2026-01-01T00:00:00Z",
            "expires_at": null,
            "source_content": null,
            "document_date": null,
            "event_dates": [],
            "importance": 0.5,
            "access_count": 0,
            "last_accessed_at": null,
            "superseded_at": null,
            "superseded_by_id": null
        })))
        .mount(&server)
        .await;

    let memory = client
        .get_memory("mem_1")
        .await
        .expect("get_memory should succeed");

    assert_eq!(memory.id, "mem_1");
}

#[tokio::test]
async fn memory_relations_returns_the_relations() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/memory/mem_1/relations"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "memory_id": "mem_1",
            "relations": []
        })))
        .mount(&server)
        .await;

    let resp = client
        .memory_relations("mem_1", None)
        .await
        .expect("memory_relations should succeed");

    assert_eq!(resp.memory_id, "mem_1");
}

#[tokio::test]
async fn forget_memory_sends_delete() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("DELETE"))
        .and(path("/v1/memory/mem_1"))
        .respond_with(ResponseTemplate::new(204))
        .mount(&server)
        .await;

    client
        .forget_memory("mem_1")
        .await
        .expect("forget_memory should succeed");
}

#[tokio::test]
async fn export_memories_returns_the_payload() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/memory/export"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "memories": [],
            "relations": []
        })))
        .mount(&server)
        .await;

    let resp = client
        .export_memories(None)
        .await
        .expect("export_memories should succeed");

    assert!(resp.get("memories").is_some());
}

#[tokio::test]
async fn import_memories_sends_memories_and_relations() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/memory/import"))
        .and(body_json(json!({
            "memories": [{"content": "x"}],
            "relations": []
        })))
        .respond_with(ResponseTemplate::new(201).set_body_json(json!({
            "source_session_id": null,
            "memories": [{
                "id": "mem_1",
                "scope": "default",
                "kind": "fact",
                "content": "x",
                "metadata": {},
                "source_session_id": null,
                "created_at": "2026-01-01T00:00:00Z",
                "updated_at": "2026-01-01T00:00:00Z",
                "expires_at": null,
                "source_content": null,
                "document_date": null,
                "event_dates": [],
                "importance": 0.5,
                "access_count": 0,
                "last_accessed_at": null,
                "superseded_at": null,
                "superseded_by_id": null
            }]
        })))
        .mount(&server)
        .await;

    let resp = client
        .import_memories(vec![json!({"content": "x"})], vec![])
        .await
        .expect("import_memories should succeed");

    assert_eq!(resp.memories.len(), 1);
    assert_eq!(resp.memories[0].id, "mem_1");
}
